src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentSynchronizer.cs

```diff
@@ -0,0 +1,166 @@
+
+            return result;
+        }
+
+        // Internal for testing
+        internal void DocumentManager_Changed(object sender, LSPDocumentChangeEventArgs args)
+        {
+            if (args is null)
+            {
+                throw new ArgumentNullException(nameof(args));
+            }
+
+            if (args.Kind != LSPDocumentChangeKind.VirtualDocumentChanged)
+            {
+                return;
+            }
+
+            var lspDocument = args.New;
+            for (var i = 0; i < lspDocument.VirtualDocuments.Count; i++)
+            {
+                if (!_synchronizingContexts.TryGetValue(lspDocument.VirtualDocuments[i].Uri, out var synchronizingContext))
+                {
+                    continue;
+                }
+
+                if (lspDocument.Version == synchronizingContext.ExpectedHostDocumentVersion)
+                {
+                    synchronizingContext.SetSynchronized(true);
+                }
+                else if (lspDocument.Version > synchronizingContext.ExpectedHostDocumentVersion)
+                {
+                    // The LSP document version has surpassed what the projected document was expecting for a version. No longer able to synchronize.
+                    synchronizingContext.SetSynchronized(false);
+                }
+            }
+        }
+
+        private class DocumentSynchronizingContext
+        {
+            private readonly TaskCompletionSource<bool> _onSynchronizedSource;
+            private readonly CancellationTokenSource _cts;
+            private bool _synchronizedSet;
+
+            public DocumentSynchronizingContext(
+                VirtualDocumentSnapshot virtualDocument,
+                int expectedHostDocumentVersion,
+                TimeSpan timeout,
+                CancellationToken requestCancellationToken)
+            {
+                VirtualDocument = virtualDocument;
+                ExpectedHostDocumentVersion = expectedHostDocumentVersion;
+                _onSynchronizedSource = new TaskCompletionSource<bool>();
+                _cts = CancellationTokenSource.CreateLinkedTokenSource(requestCancellationToken);
+
+                // This cancellation token is the one passed in from the call-site that needs to synchronize an LSP document with a virtual document.
+                // Meaning, if the outer token is cancelled we need to fail to synchronize.
+                _cts.Token.Register(() => SetSynchronized(false));
+                _cts.CancelAfter(timeout);
+            }
+
+            public VirtualDocumentSnapshot VirtualDocument { get; }
+
+            public int ExpectedHostDocumentVersion { get; }
+
+            public Task<bool> OnSynchronizedAsync => _onSynchronizedSource.Task;
+
+            public void SetSynchronized(bool result)
+            {
+                lock (_onSynchronizedSource)
+                {
+                    if (_synchronizedSet)
```

Comment:
Is this check here so that if it is already set as true, then it won't be set as false when the cts is cancelled?

Reply:
You can only set the result once. Any other settings (even to the same value) will throw